move to documentation website

jaxxzer opened the pull request:
Note actual commit topology in this pr is very different from github display order, it might help to do the review in `tig`.

jaxxzer commented:
Documentation website preview: http://docs.bluerobotics.com/ping-protocol/

jaxxzer commented:
@patrickelectric @Williangalvani I can't figure out why travis stopped running, I need help. Feel free to force push.

jaxxzer commented:
resolved, thanks.

jaxxzer commented:
Next steps:
- Check out the new [ping-cpp](https://github.com/bluerobotics/ping-cpp) repository (including the [deployment](https://github.com/bluerobotics/ping-cpp/tree/deployment) branch)
- review the ping-viewer pr to move to ping-cpp submodule:
- I will soon submit a ping-arduino pr to use ping-cpp submodule in the same manner as ping-viewer ( I will make arduino work with the submodule, don't worry about that)
- I will soon submit a ping-python pr to migrate the test/ci/deployment that was inside of the ping-protocol repository into the ping-python repository (make ping-python have ci and structure that is similar to ping-cpp repository)

jaxxzer commented:
Here is an image that describes the repository layout I envision at the end of this work: 
![ping-protocol-planning - New frame (2)](https://user-images.githubusercontent.com/8315108/59702450-bc955980-91c5-11e9-9a8d-b4b07a580cb2.jpg)

jaxxzer commented:
I've recorded a few advantages we will see with this work:
- ready-to-checkout source code is available in implementation repositories after checkout of deployment branch (no need to run python locally) (ping viewer no longer depends on python)
- we now have an independent minimal cpp library, that has no dependencies
- other dependency reductions, simplified ci scripts, and faster ci times
- NO AUTOGENERATED content is tracked in development branch (no more forgetting to update protocol readme and failing ci)
- arduino and ping viewer share parser code (both previous parsers had bugs!)
- generator class is improved with generate method, and generation scripts and templates are reduced and simplified
- development and deployment is appropriately compartmentalized:
  - deployment of implementation repositories is independent of commits in ping-protocol (no more broken implementations)
  - python files moved to python repo, cpp files moved to cpp repo, development of implementation apis is reflected in implementation repository history (instead of remote ping-protocol repository)
- test cases are made by hand instead of generated (more robust):
  - static test cases (not autogenerated) mean that if we break the api, we will know (ex. when refactoring class name PingMessage -> ping_message)
  - automatically generated test cases risk bugs in the implementation (bug in the code generating the test case may cause actual problems to go undetected)

patrickelectric commented on .travis.yml:
```diff
@@ -1,73 +1,26 @@
-- if [ "${PINGTEST}" = "true" ]; then
-    tools/run-tests.sh || travis_terminate 1;
-    tools/update-generated-repositories.sh;
+- |
```
I think this should be in a script inside like the old version, there is no reason to be here.
Add this kind of stuff inside yaml makes it harder to maintain and test.

jaxxzer replied:
Agreed.

patrickelectric commented on src/protocol/generate-markdown.py:
```diff
@@ -0,0 +1,32 @@
+
+templateFile = "%s/pingmessage-.md.in" % templatePath
+
+g = Generator()
```
Do not use single letter variables

patrickelectric commented on src/protocol/generator.py:
```diff
@@ -1,17 +1,14 @@
 
-from jinja2 import Environment, FileSystemLoader
-
+# TODO the generator api should help with determining things about messages (like a total payload length, or if a field is a vector type)
```
Avoid long lines with comments, break it in more lines.

patrickelectric replied:
Use docstrings: https://pythonhelp.wordpress.com/2011/02/14/docstrings/

patrickelectric commented on src/protocol/generator.py:
```diff
@@ -52,7 +49,7 @@ def get_c_size(self, t):
         match = re.search('[0-9]{1,2}', t)
         if match:
-            return int(match.group(0)) / 8
+            return int(int(match.group(0)) / 8)
```
why is this necessary ?

jaxxzer replied:
Previously, only the dividend was converted, and the return was still float. My fix here was not the best (just needed to move a parenthesis). I'll take care of it.

jaxxzer replied:
(We were getting things like `msgData[2.0]`, this patch fixes that)

patrickelectric commented on src/protocol/generator.py:
```diff
@@ -101,44 +98,25 @@ def get_type_string(self, t, pointer=False, name=''):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globies=None):
+        d = json.load(open(definition), object_pairs_hook=collections.OrderedDict)
```
Avoid single letter variables.

patrickelectric commented on src/protocol/templates/pingmessage_.h.in:
```diff
@@ -9,47 +9,48 @@
-{% set total_payload = generator.calc_payload(messages[message_type]['all_msgs'][message].payload) %}
+{% for message_type in messages %}
+{% for message in messages[message_type] %}
+{% set m = messages[message_type][message] %}
```
Avoid single letter variables

jaxxzer replied:
(this file has been removed, I checked the others)

jaxxzer commented:
Did a few more things this morning:
- removed the stale templates
- moved `src/protocol/*` to `just src/*`
- updated gitignore
now the repository is totally clean: https://github.com/jaxxzer/ping-protocol/tree/doc

jaxxzer commented:
@patrickelectric done

patrickelectric commented on tools/travis-ci-script.sh:
```diff
@@ -0,0 +1,21 @@
+    echo "Json file does not follow style. Try python3 -m json.tool."
+    exit 1
+fi
+done
```
It's missing the indentation here.

Williangalvani commented on src/generator.py:
```diff
@@ -101,44 +102,26 @@ def get_type_string(self, t, pointer=False, name=''):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globuls=None):
+        definitionDict = json.load(open(definition), object_pairs_hook=collections.OrderedDict)
```
while it will probably work fine, using `open()` like this leaves the file opened (ppparently until is is garbage-collected).

I'd rather use the [`with open(..)`](https://docs.python-guide.org/writing/style/#read-from-a-file) syntax or at least explicitly close the file.

jaxxzer replied:
done

Williangalvani commented on src/generator.py:
```diff
@@ -1,17 +1,17 @@
-from jinja2 import Environment, FileSystemLoader
+import collections
+from jinja2 import Template
+from pathlib import Path
```
Is this used?

jaxxzer replied:
nope!

Williangalvani commented on src/generator.py:
```diff
@@ -101,44 +102,26 @@ def get_type_string(self, t, pointer=False, name=''):
-    def capitalize(self, input_string):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globuls=None):
```
What is a "globul"? should it be "glob"?

jaxxzer replied:
`global` is a python keyword

Williangalvani commented on src/generator.py:
```diff
@@ -52,7 +53,7 @@ def get_c_size(self, t):
         match = re.search('[0-9]{1,2}', t)
         if match:
-            return int(match.group(0)) / 8
+            return int(match.group(0) / 8)
```
Does it work?
I expected the result of  `match.group()` to be a string, which should not be divisible by 8.

jaxxzer replied:
it does with python3 :) and travis ci

jaxxzer replied:
> I expected the result of match.group() to be a string, which should not be divisible by 8.

though, that is probably why I found it the way it was. We (I) started development of this repo with python 2.

patrickelectric commented on tools/travis-ci-script.sh:
```diff
@@ -0,0 +1,21 @@
+
+for filename in src/definitions/*.json; do
+echo "Checking file: ${filename}"
+python3 -m json.tool ${filename} > /tmp/temporary_test_file.json;
```
still missing indentation.

jaxxzer replied:
XXXXXXXX
